src/data-structures/Hashmap.balls

```diff
@@ -0,0 +1,100 @@
+/// @title HashMap
+/// @notice SPDX-License-Identifier: MIT
+/// @author asnared <https://github.com/abigger87>
+
+/// @notice A Module Encapsulating HashMap Methods
+/// @notice Adapted from <https://github.com/huff-language/huff-examples/blob/main/erc20/contracts/utils/HashMap.huff>
+
+/// @notice Given a piece of data (ie an address), hash it, generating the storage slot for a hashmap.
+/// @notice Port to BALLs by eugenioclrc taken from https://github.com/huff-language/huffmate/blob/4e2c9bd3412ab8cc65f6ceadafc01a1ff1815796/src/data-structures/Hashmap.huff
+
+/// @notice Given a piece of data (ie an address), hash it, generating the storage slot for a hashmap.
+fn GET_SLOT_FROM_KEY<mem_ptr>(key) -> (hashed) {
+    // Load the data into memory and hash it, while preserving the memory location.
+    mstore(mem_ptr, key)
+    // Hash the data, generating a key.
+    hashed = sha3(mem_ptr, 0x20)
+}
+
+/// @notice Given two keys (ie a slot and a key), hash them together, generating a slot for a secondary hashmap.
+fn GET_SLOT_FROM_KEYS<mem_ptr>(slot, key) -> (hashed) {
+    // Load the data into memory.
+    mstore(add(mem_ptr, 0x20), slot)
+    mstore(mem_ptr, key)
+    
+    
+    // Hash the data, generating a slot.
+    hashed = sha3(mem_ptr, 0x40)
+}
+
+
+/// @notice Calculate the slot from two keys
+fn GET_SLOT_FROM_KEYS_2D<mem_ptr>(slot, key1, key2) -> (hashed) {
+    // Load the data into memory.
+    mstore(add(mem_ptr, 0x20), slot)
+    mstore(mem_ptr, key1)
+    sha3(mem_ptr, 0x40)
```

Comment:
the result of this hash needs to be stored and set into memory. Also wonder if you could save gas by storing `add(mem_ptr, 0x20)` in a var and just referencing that again